[1단계 - 웹 자동차 경주] 아마란스(송세연) 미션 제출합니다.

amaran-th opened the pull request:
안녕하세요 춘식! 이번 미션 리뷰이가 된 아마란스라고 합니다ㅎㅎ
스프링 프레임워크를 거의 써본 적이 없어서 단순한 요구사항임에도 애를 먹었던 것 같아요
개선할 수 있을 부분이 보이신다면 리뷰 부탁드립니다
감사합니다!

BackFoxx committed:
feat: 레벨 1에서 사용한 자동차경주 구현코드 적용

BackFoxx committed:
docs: 기능 요구사항 정리

BackFoxx committed:
docs: 데이터베이스 테이블 DDL 작성

BackFoxx committed:
build: 데이터베이스 관련 의존성 추가 및 설정

BackFoxx committed:
feat: HTTP API 요청/응답 구현 - 여우

BackFoxx committed:
feat: HTTP API 요청/응답 구현 - 아마란스

BackFoxx committed:
feat: 잘못된 요청이 들어왔을 때 예외 처리 - 여우

BackFoxx committed:
feat: 잘못된 요청이 들어왔을 때 예외 처리 - 아마란스

BackFoxx committed:
feat: 자동차경주와 관련된 정보를 데이터베이스에 insert 하는 기능 구현 - 아마란스

BackFoxx committed:
docs: 구현사항을 README에 반영

bimppap commented on src/main/java/racingcar/dao/UpdatingDAO.java:
```diff
@@ -0,0 +1,43 @@
+import java.sql.PreparedStatement;
+
+@Repository
+public class UpdatingDAO {
```
Repository와 DAO의 차이는 무엇일까요?

bimppap replied:
클래스명은 UpdatoinDAO인데 실제로 돌아가는 쿼리는 insert네요?

amaran-th replied:
지금까지 Repository와 DAO가 동일한 개념이라고 생각했었는데, 찾아보니 Repository는 엔티티 객체를 보관하고 관리하는 저장소, DAO는 데이터에 접근할 수 있도록 DB 접근 관련 로직을 모아둔 객체라고 하네요.

미션 전에 제공된 학습 테스트 코드에서 조회 기능의 경우 QueryingDAO 클래스에서 구현했고 그 외 insert, delete, update 기능은 UpdatingDAO 클래스에서 구현했는데 거기에서 가져왔습니다! 혼동할 수 있는 네이밍인 것 같아 InsertingDAO로 변경해주었습니다!

찾아보니 @Repository를 Repository 구현체와 DAO 클래스에 모두 사용하는 것 같아서 어노테이션을 지워주진 않았는데, 제가 이해한 게 맞을까요??

bimppap replied:
아마란스 말대로 `@Repository` 어노테이션은 repository 와 dao 모두 쓸 수 있습니다. 잘 찾아보셨네요. 👍 

좀 더 쉽게 얘기해보자면 DAO는 DB와 가까운 개념이고 Repository는 도메인에 가까운 개념입니다. 만약 둘이 같이 있다면, DAO는 레포지토리를 사용할 수 없지만 Repository는 DAO를 사용할 수 있습니다. 하지만 도메인의 크기가 작은 경우에는 둘을 동일하다 봐도 무방하기 때문에 지금 분리해볼 필요는 없습니다. 🙂 (저는 개인적으로 [이 글](https://m.blog.naver.com/PostView.naver?isHttpsRedirect=true&blogId=loopbit&logNo=221216795623)을 재밋게 읽었습니다ㅎㅎ)

그리고 지금은 insert 쿼리밖에 없어서 InsertingDAO라고 해도 괜찮지만, 보통은 `[사용하는 도메인 또는 업무]DAO/Repository` (예시: CarDAO) 라고 지으니 참고바랍니다ㅎㅎ

bimppap commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -0,0 +1,116 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Controller
```
RestController 와 Controller의 차이는 무엇일까요?

amaran-th replied:
잘 모르고 있었던 어노테이션이라 찾아보고 정리해보았습니다!

- `@Controller` :  클라이언트의 요청에 대해 View를 반환하기 위해 사용한다.
    - 컨트롤러가 반환한 뷰의 이름으로 view 파일(일반적으로 html)를 찾아 렌더링한다.
- `@RestController` : view가 아니라 Data를 반환할 때, 즉 RestAPI의 응답을 보내기 위해 사용한다.
    
    메서드 차원에서 `@ResponseBody` 어노테이션을 사용한 것과 동일한 기능을 한다.

메서드에 일일이 @ResponseBody를 붙여주는 것보다 컴포넌트 차원에서 @RestController를 붙여주는 게 좋을 것 같아 수정하였습니다~~!
    

[참고 게시글](https://mangkyu.tistory.com/49)

bimppap commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -0,0 +1,116 @@
+        validateDuplication(carNames);
+        return saveCars(carNames);
+    }
+
```
컨트롤러에 이 메소드가 꼭 있어야할 이유가 있을까요? Cars에 메세지를 보내는 방식으로 하면 어떨까요? 추천한 방법으로 변경하면 어떤 장점이 있을지 고민해보면 좋을 것 같습니다 🙂

amaran-th replied:
반영하였습니다!

bimppap commented on src/main/java/racingcar/vo/CarName.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.List;
+import java.util.Objects;
+
+public class CarName {
```
VO와 Domain을 분리한 이유가 있을까요?

amaran-th replied:
Domain 패키지에 포함시켜주었습니다.

bimppap commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,46 @@
+
+    private Car(CarName name) {
+        this.name = name;
+        this.position = Position.of(0);
```
Position에서 ZERO_POSITION 이라는 이름의 static 변수가 있어도 좋을 것 같습니다

amaran-th replied:
반영해주었습니다!

bimppap commented on src/main/java/racingcar/dto/RequestBodyDTO.java:
```diff
@@ -0,0 +1,14 @@
+package racingcar.dto;
+
+public class RequestBodyDTO {
```
이름만 봐서는 어디서 쓰이는지 잘 모르겠는데, 어떻게 생각하시나요?

amaran-th replied:
RacingDTO로 수정했습니다!

bimppap commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,4 @@
+spring.h2.console.enabled=true
```
properties 와 yml의 차이는 무엇이 있을까요?

amaran-th replied:
찾아보니 문법적인 차이가 있네요! 계층 구조를 표현하기엔 yml이 더 가독성이 좋다고 하는데, 개인적으로 yaml 문서를 작성하다가 들여쓰기 때문에 고생한 적이 있어서 선호하지 않는 편입니다...ㅎ
그래서 이번 미션에서는 따로 수정하진 않겠습니다!

bimppap commented on src/main/resources/static/index.html:
```diff
@@ -50,21 +50,25 @@ <h1>자동차 경주 게임</h1>
+                      document.getElementById("resultPosition").innerHTML = '';
+                      document.getElementById("resultPosition").appendChild(resultPosition);
+          } else {
+            window.alert("Error:"+data.message);
```
띄워쓰기가 필요할 것 같군요 😎코드 포맷팅을 습관화 합시다!

amaran-th replied:
자바 코드는 자동 포매팅이 되도록 설정했는데 html은 신경써주지 못했네요...! 수정했습니다.

bimppap commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,99 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class CarsTest {
```
실패하는 테스트가 있는데 확인 부탁드립니다! 커밋 전에 테스트를 한번씩 돌려보는 습관을 들이면 좋습니다 🙂

amaran-th replied:
모든 테스트가 통과되도록 수정했습니다!

bimppap commented on src/main/resources/data.sql:
```diff
@@ -1,7 +1,16 @@
-    PRIMARY KEY (id)
+CREATE TABLE racing(
+    `id` INT NOT NULL AUTO_INCREMENT,
+    `trialCount` INT NOT NULL,
+    `created_at` DATETIME NOT NULL DEFAULT current_timestamp,
+    PRIMARY KEY(id)
```
케이스 스타일은 통일시켜 주세요!

amaran-th replied:
수정했습니다!

bimppap commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @ResponseBody
+    @ExceptionHandler
```
IllegalArgumentException이 아닌 에러가 오면 어떻게 되나요? 👀

bimppap replied:
ExceptionHandler와 ControllerAdvice에 대해서도 알아보면 좋을 것 같습니다

amaran-th replied:
사용자에 의한 오류(IllegalArgumentException)의 경우 프론트에서 alert를 띄울 수 있게 해뒀는데, 그 외 오류들도 동일하게 처리할 수 있도록 해주는 게 더 안정적일 것 같네요!
프론트에서 따로 처리해줄 수 있도록 오류 별로 type 속성을 다르게 해서 응답을 던지도록 수정했습니다.

ControllerAdvice는 처음 알았는데 에러 처리 클래스를 따로 분리해줄 수 있을 것 같아서 좋아보이네요! 마찬가지로 코드에 반영했습니다.

bimppap commented on src/main/java/racingcar/dao/UpdatingDAO.java:
```diff
@@ -0,0 +1,43 @@
+        String sql = "INSERT INTO racing (trialCount) values (?)";
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        jdbcTemplate.update(
+                con -> {
```
con 보다는 connection을 쓰는게 어떨까요?

amaran-th replied:
반영하였습니다!

bimppap commented on src/main/java/racingcar/dao/UpdatingDAO.java:
```diff
@@ -0,0 +1,43 @@
+        return keyHolder.getKey().intValue();
+    }
+
+    public void insert(CarInfoDto carInfoDto) {
```
위 메서드는 PreparedStatement를 쓰고 여기선 안 썼는데 이유가 있을까요?

amaran-th replied:
racing 데이터 insert와 car_info 데이터 insert 모두 학습 테스트 코드를 참고해서 작성했는데, 그 코드에서 KeyHolder를 활용하는 코드에서만 PreparedStatement를 사용해서 그런 것 같습니다!
preparedStatement를 사용하는 게 여러 면에서 좋을 것 같아 KeyHolder를 사용하지 않는 메서드에도 PreparedStatement를 적용하였습니다.

bimppap replied:
PreparedStatement은 보안, 성능적인 장점도 있지만 단점도 있으니 나중에 한번 장단점을 정리해보면 좋을 것 같습니다 🙂

bimppap commented on build.gradle:
```diff
@@ -12,6 +12,8 @@ repositories {
 
 dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-web'
+	implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+	runtimeOnly 'com.h2database:h2'
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
```
implementation, api, compileOnly, runtimeOnly 의 차이점은 무엇일까요?

amaran-th replied:
의존성이라는 개념이 아직 명확히 이해되진 않지만 제가 이해한 부분까지 정리해보았습니다!

- `ClassPath` : 클래스나 jar 파일이 존재하는 위치
    - `Compile-time classpath` : 에러 없이 컴파일을 하기 위해 필요한 클래스와 jar들의 위치
    - `Run-time classpath` : 애플리케이션이 정상적으로 실행되기 위해 필요한 클래스들과 jar들의 위치

```java
class A{
  public static void main(String[] args){
    B b = new B();
  }
}

class B {
  public B() {
    return new C().sum();
  }
}

class C {
  public int sum(){
    return 5;
  }
}
```

위와 같은 코드가 있을 때, 의존성 방향은 `A→B→C`로 흐른다.

여기서 컴파일 시점에서는 A와 B만 알고 있으면 오류가 발생하지 않으나, 런타임을 위해서는 C에 대한 정보도 가지고 있어야 한다. 이를 A와 B가 컴파일 타임 의존성을 가지고, A, B, C가 런타임 의존성을 가지고 있다고 할 수 있다.

- `implementation` : 의존성을 compile classPath, runtime classPath 두 곳에 모두 추가한다. 라이브러리가 수정될 경우 직접적으로 의존관계에 있는 라이브러리만 rebuild된다.
- `testImplementation` : 테스트 코드를 수행할 때만 적용.
- `api` : implementation과 유사하나, 라이브러리가 수정이 될 때 해당 라이브러리를 의존하고 있는 다른 라이브러리들도 모두 rebuild가 일어난다는 특징이 있다.
    - 속도 면에서 문제가 있다.
- `compileOnly` : 의존성을 compileClassPath에만 추가한다.
    - 빌드 결과물의 사이즈가 줄어든다.
- `runtimeOnly` : 의존성을 runtimeClassPath에만 추가한다.
    - 해당 클래스에서 코드 변경이 발생해도 컴파일을 다시 할 필요가 없다.
    
*혹시 잘못된 정보가 있다면 알려주시면 감사하겠습니다!

bimppap replied:
잘 정리해주셨네요ㅎㅎ 고생 많으셨습니다 🥰

bimppap commented on src/main/java/racingcar/utils/Converter.java:
```diff
@@ -0,0 +1,17 @@
+
+    public static int convertStringToLong(String target) {
+       try {
+           return Integer.parseInt(target);
```
Long으로 컨버팅하는 메서드인데 리턴타입은 int네요?

amaran-th replied:
Long 타입을 Int로 바꿔주는 과정에서 누락된 것 같네요! 수정했습니다.

bimppap commented on src/main/java/racingcar/dto/RacingCarsDto.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.function.BiConsumer;
+
+public class RacingCarsDto {
+    private final Map<String, Integer> result;
```
Dto가 map 을 가지고 있어도 될까요?

amaran-th replied:
생각해보니 RacingCarDto는 콘솔 프로그램에서 사용되는 DTO인데, 콘솔 프로그램에서 굳이 DTO를 사용할 필요는 없을 것 같아 삭제했습니다!

bimppap commented on src/main/java/racingcar/dto/RacingCarsDto.java:
```diff
@@ -0,0 +1,28 @@
+        return new RacingCarsDto(resultHolder);
+    }
+
+    public void forEach(BiConsumer<? super String, ? super Integer> action) {
```
dto는 계층간 데이터 교환을 위한 객체인데 로직이 있어도 될까요?

amaran-th replied:
굳이 CarDto를 한 번 더 감싸는 것보다 List<CarDto>(콘솔 프로그램에서는 List<Car>)로 처리해주는 편이 나을 것 같아 그렇게 수정했습니다.

bimppap commented on docs/README.md:
````diff
@@ -0,0 +1,160 @@
+
+```
+
+# 💠클래스 다이어그램
````
리드미를 꼼꼼하게 짜셨네요! 👍
클래스 다이어그램은 2단계 때 작성 예정인건가요?ㅎㅎ

bimppap replied:
루트 경로에도 README.md가 있어서, 하나로 통일해주면 좋을 것 같습니다🙂

bimppap replied:
gitkeep 파일이 두 개 있는데 이유가 있을까요? 아니라면 파일 정리는 미리미리 해주는게 좋습니다!

amaran-th replied:
1. 네! 경험 상 클래스 구조는 개발/리팩토링 과정에서 계속 변화해서, 제 때 수정사항을 반영하기가 번거롭더라구요. 클래스 다이어그램은 어느정도 마무리되고 난 뒤에 작성할 예정입니다!

2. 제가 우테코 미션을 진행할때 개인적으로 세운 규칙이 하나 있는데, docs 폴더의 README는 미션 과정에서 필요한 정보를 기록하고 변경사항은 즉각적으로 반영하는 메모의 목적으로 작성하고, 미션이 끝난 뒤에 내용을 다듬어서 최상위 경로에 있는 README에 작성해주고 있습니다!

3. 1레벨의 자동차 경주 코드를 가져오면서 중복된 것 같습니다! 미처 확인하지 못하고 넘어간 것 같네요😅
현재 코드에선 필요하지 않을 것 같아 삭제했습니다! 알려주셔서 감사합니다!

bimppap commented on src/main/java/racingcar/RacingCarApplication.java:
```diff
@@ -5,9 +5,7 @@
 public class RacingCarApplication {
-
 	public static void main(String[] args) {
 		SpringApplication.run(RacingCarApplication.class, args);
```
잘 동작하네요 👍
<img width="229" alt="image" src="https://user-images.githubusercontent.com/48675973/231493312-dc119502-b0ed-4a02-9341-cc6c1a26ea56.png">

bimppap reviewed (changes requested):
안녕하세요 아마란스, 춘식입니다!
방학은 잘 보내셨나요? 새로 시작하는 레벨 2도 화이팅하시길 바랍니다ㅎㅎ
1단계 미션 잘 구현해주셨네요. 코멘트를 조금 남겨봤으니 확인 부탁드립니다!
궁금하거나 이해가 안 되는 점은 편하게 DM 주세요 😎

Commit:
feat : 불필요한 .gitkeep 파일 제거

Commit:
feat : WebController의 어노테이션 @Controller를 @RestController로 변경

Commit:
feat : 1레벨 코드 대폭 수정

- inputView의 검증 코드 제거 및 객체 생성 코드를 컨트롤러로 이동
- 테스트 코드 컴파일 오류 수정
- 띄어쓰기 컨벤션 반영

Commit:
feat : 에러 처리 코드 수정

- @RestContorlloerAdvice 어노테이션으로 에러 처리 클래스 분리
- ExceptionMessageDTO에 type 속성 추가
- 프론트 코드 수정

Commit:
refactor : vo 패키지를 domain 내부로 이동

Commit:
refactor : UpdatingDAO를 InsertingDAO로 이름 변경

Commit:
feat : insertDAO에 preparedStatement 적용

Commit:
refactor : 자동차 초기 위치(position) 상수화

Commit:
refactor : Converter 메서드 명칭 수정 및 데이터베이스 컬럼명 컨벤션 통일

Commit:
refactor : 코드 컨벤션 수정

amaran-th reviewed (commented):
안녕하세요 춘식!
고민해보면 좋을 점까지 정성스레 리뷰해주셔서 정말 감사해요:)
코멘트에도 남겨놨듯 제가 스프링에서 어떤 걸 모르는지도 잘 모르는 상태였는데 덕분에 많이 알게 된 것 같아요! 
최대한 이해한 대로 수정하고 반영해보았습니다.
시간 나실 때 리뷰 부탁드려요ㅎㅎ

Commit:
feat : RacingCarsDto 제거

Commit:
style : 코드 컨벤션 맞추기

bimppap commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -0,0 +1,51 @@
+@RestController
+public class WebController {
+
+    final private InsertingDAO insertingDAO;
```
2단계에서는 Service에 대해 알아보고, 이를 이용해 컨트롤러의 책임을 좀 더 작게 만들어볼까요?

bimppap reviewed (approved):
안녕하세요 아마란스, 춘식입니다!
리팩토링 정말 잘해주셨어요. 많이 찾아보고 공부한 흔적이 보이네요 👍
이번 단계는 여기서 머지하도록 하겠습니다. 수고 많으셨어요.
질문에 대한 코멘트 남겼으니 확인 바랍니다.
궁금하거나 이해가 안 되는 점이 있으면 편하게 DM 주세요 🙂
그럼 다음 단계에서 만나요!
